### PR TITLE
Add usub_cond and usub_sat operations to atomicrmw

### PR DESCRIPTION
**Pull request by anjenner**
These both perform conditional subtraction, returning the minuend and zero respectively, if the difference is negative.

### CODE REVIEW

File: llvm/docs/LangRef.rst [hunk 1]
```diff
@@ -11291,6 +11293,8 @@ operation argument:
 -  fmin: ``*ptr = minnum(*ptr, val)`` (match the `llvm.minnum.*`` intrinsic)
 -  uinc_wrap: ``*ptr = (*ptr u>= val) ? 0 : (*ptr + 1)`` (increment value with wraparound to zero when incremented above input value)
 -  udec_wrap: ``*ptr = ((*ptr == 0) || (*ptr u> val)) ? val : (*ptr - 1)`` (decrement with wraparound to input value when decremented below zero).
+-  usub_cond: ``*ptr = (*ptr u>= val) ? *ptr - val : *ptr`` (subtract only if no unsigned overflow).
+-  usub_sat: ``*ptr = (*ptr u>= val) ? *ptr - val : 0`` (subtract with clamping to zero).
```
Comment (jayfoad) [inline llvm/docs/LangRef.rst:11297]: Nit: maybe "with unsigned clamping to zero" would be more rigorous.

File: llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h [hunk 2]
```diff
@@ -1636,6 +1636,41 @@ class MachineIRBuilder {
...
+  MachineInstrBuilder buildAtomicRMWUSubCond(const DstOp &OldValRes,
+                                             const SrcOp &Addr,
+                                             const SrcOp &Val,
+                                             MachineMemOperand &MMO);
+
+  /// Build and insert `OldValRes<def> = G_ATOMICRMW_USUB_SAT Addr, Val, MMO`.
+  ///
+  /// Atomically replace the value at \p Addr with the original value minus \p
+  /// Val if the original value is greater than or equal to \p Val, or with zero
+  /// otherwise. Puts the original value from \p Addr in \p OldValRes.
```
Comment (jayfoad) [inline llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h:1661]: Nit: maybe explain this in a more user-friendly way: "... with the original value minus Val, with clamping to zero if the unsigned subtraction would overflow."

File: llvm/lib/Transforms/Utils/LowerAtomic.cpp [hunk 3]
```diff
@@ -95,6 +95,17 @@ Value *llvm::buildAtomicRMWValue(AtomicRMWInst::BinOp Op,
     Value *Or = Builder.CreateOr(CmpEq0, CmpOldGtVal);
     return Builder.CreateSelect(Or, Val, Dec, "new");
   }
+  case AtomicRMWInst::USubCond: {
+    Value *Cmp = Builder.CreateICmpUGE(Loaded, Val);
+    Value *Sub = Builder.CreateSub(Loaded, Val);
+    return Builder.CreateSelect(Cmp, Sub, Val, "new");
```
Comment (jayfoad) [inline llvm/lib/Transforms/Utils/LowerAtomic.cpp:101]: Shouldn't this be:
```suggestion
    return Builder.CreateSelect(Cmp, Sub, Loaded, "new");
```
?

File: llvm/lib/Transforms/Utils/LowerAtomic.cpp [hunk 4]
```diff
@@ -95,6 +95,17 @@ Value *llvm::buildAtomicRMWValue(AtomicRMWInst::BinOp Op,
...
+  }
+  case AtomicRMWInst::USubSat: {
+    Constant *Zero = ConstantInt::get(Loaded->getType(), 0);
+    Value *Cmp = Builder.CreateICmpUGE(Loaded, Val);
+    Value *Sub = Builder.CreateSub(Loaded, Val);
+    return Builder.CreateSelect(Cmp, Sub, Zero, "new");
```
Comment (jayfoad) [inline llvm/lib/Transforms/Utils/LowerAtomic.cpp:107]: Could you do this instead?
```suggestion
    return Builder.CreateIntrinsic(Intrinsic::usub_sat, Loaded->getType(), {Loaded, Val}, nullptr, "new");
```

File: llvm/bindings/ocaml/llvm/llvm.ml [hunk 5]
```diff
@@ -296,6 +296,12 @@ module AtomicRMWBinOp = struct
   | UMin
   | FAdd
   | FSub
+  | FMax
+  | FMin
+  | UInc_Wrap
+  | UDec_Wrap
```
Comment (RKSimon) [inline llvm/bindings/ocaml/llvm/llvm.ml:302]: Should the various FMax/FMin/UInc_Wrap/UDec_Wrap additions be a separate PATCH?

File: llvm/bindings/ocaml/llvm/llvm.mli [hunk 6]
```diff
@@ -331,6 +331,12 @@ module AtomicRMWBinOp : sig
   | UMin
   | FAdd
   | FSub
+  | FMax
+  | FMin
+  | UInc_Wrap
```
Comment (arsenm) [inline llvm/bindings/ocaml/llvm/llvm.mli:336]: Same here

File: llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h (hunk 2 above)
Comment (arsenm) [reply llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h:1661]: I don't think it's useful wasting breath on documenting the semantics of these operations here. There is dedicated G_* opcode documentation, but even that is redundant with the LangRef

File: llvm/bindings/ocaml/llvm/llvm.ml (hunk 5 above)
Comment (anjenner) [reply llvm/bindings/ocaml/llvm/llvm.ml:302]: I have created https://github.com/llvm/llvm-project/pull/105673 to add the previously missing bindings.

File: llvm/lib/Transforms/Utils/LowerAtomic.cpp [hunk 7]
```diff
@@ -98,13 +98,15 @@ Value *llvm::buildAtomicRMWValue(AtomicRMWInst::BinOp Op,
...
-    Constant *Zero = ConstantInt::get(Loaded->getType(), 0);
-    Value *Cmp = Builder.CreateICmpUGE(Loaded, Val);
-    Value *Sub = Builder.CreateSub(Loaded, Val);
-    return Builder.CreateSelect(Cmp, Sub, Zero, "new");
+    return Builder.CreateIntrinsic(Intrinsic::usub_sat, Loaded->getType(),
+                                   {Loaded, Val}, nullptr, "new");
+    //    Constant *Zero = ConstantInt::get(Loaded->getType(), 0);
+    //    Value *Cmp = Builder.CreateICmpUGE(Loaded, Val);
+    //    Value *Sub = Builder.CreateSub(Loaded, Val);
+    //    return Builder.CreateSelect(Cmp, Sub, Zero, "new");
```
Comment (arsenm) [inline llvm/lib/Transforms/Utils/LowerAtomic.cpp:109]: Remove commented out code

File: llvm/lib/Transforms/InstCombine/InstCombineAtomicRMW.cpp [hunk 8]
```diff
@@ -41,6 +41,8 @@ bool isIdempotentRMW(AtomicRMWInst& RMWI) {
     case AtomicRMWInst::Sub:
     case AtomicRMWInst::Or:
     case AtomicRMWInst::Xor:
+    case AtomicRMWInst::USubCond:
+    case AtomicRMWInst::USubSat:
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineAtomicRMW.cpp:45]: Missing tests for this (I would also split the instcombine piece into a separate change)

File: llvm/lib/CodeGen/AtomicExpandPass.cpp [hunk 9]
```diff
@@ -1562,6 +1564,8 @@ bool AtomicExpandImpl::isIdempotentRMW(AtomicRMWInst *RMWI) {
   case AtomicRMWInst::Sub:
   case AtomicRMWInst::Or:
   case AtomicRMWInst::Xor:
+  case AtomicRMWInst::USubCond:
+  case AtomicRMWInst::USubSat:
```
Comment (arsenm) [inline llvm/lib/CodeGen/AtomicExpandPass.cpp:1568]: Missing tests for this (but this optimization is likely to be deleted soon anyway). Missing any new tests in test/Transforms/AtomicExpand

File: llvm/lib/Transforms/InstCombine/InstCombineAtomicRMW.cpp (hunk 8 above)
Comment (anjenner) [reply llvm/lib/Transforms/InstCombine/InstCombineAtomicRMW.cpp]: I've removed this piece for now. I'm happy to add it back with a separate change, but it looks from the tests like we don't actually want to use idempotency optimizations with atomicrmw operations at all so maybe nothing needs to be done here.

File: llvm/lib/CodeGen/AtomicExpandPass.cpp (hunk 9 above)
Comment (anjenner) [reply llvm/lib/CodeGen/AtomicExpandPass.cpp]: If the optimization is likely to be deleted soon, then removing this part of the change seems like a better idea than adding new tests for it, so I have done that.

File: llvm/lib/Transforms/InstCombine/InstCombineAtomicRMW.cpp (hunk 8 above)
Comment (arsenm) [reply llvm/lib/Transforms/InstCombine/InstCombineAtomicRMW.cpp]: The atomicrmw -> atomicrmw idempotent case handled here is fine, it's the atomicrmw -> load atomic case that's problematic in the lowering.

Also in a follow up, the AMDGPU lowering should have the idempotent case to add 0 since that works over PCIe

File: llvm/lib/CodeGen/AtomicExpandPass.cpp [hunk 10]
```diff
@@ -1803,6 +1805,8 @@ static ArrayRef<RTLIB::Libcall> GetRMWLibcall(AtomicRMWInst::BinOp Op) {
   case AtomicRMWInst::FSub:
   case AtomicRMWInst::UIncWrap:
   case AtomicRMWInst::UDecWrap:
+  case AtomicRMWInst::USubCond:
+  case AtomicRMWInst::USubSat:
     // No atomic libcalls are available for max/min/umax/umin.
```
Comment (arsenm) [inline llvm/lib/CodeGen/AtomicExpandPass.cpp:1810]: Comment was already out of date, but it probably shouldn't bother listing out all the names of the operations

File: llvm/include/llvm/IR/Instructions.h [hunk 11]
```diff
@@ -751,8 +751,16 @@ class AtomicRMWInst : public Instruction {
     /// *p = ((old == 0) || (old u> v)) ? v : (old - 1)
     UDecWrap,
 
+    /// Subtract only if result would be positive.
+    /// *p = (old u>= v) ? old - v : old
+    USubCond,
+
+    /// Subtract with clamping of negative results to zero.
+    /// *p = (old u>= v) ? old - v : 0
```
Comment (arsenm) [inline llvm/include/llvm/IR/Instructions.h:759]: Comment in terms of the intrinsic? Also could update the phrasing to match langref

**Review by arsenm — APPROVED** [review]

**Comment by kazutakahirata** [comment]
This patch seems to trigger:

```
llvm/lib/Target/AMDGPU/AMDGPULowerBufferFatPointers.cpp:1101:13: error: enumeration values 'USubCond' and 'USubSat' not handled in switch [-Werror,-Wswitch]
    switch (RMW->getOperation()) {
            ^~~~~~~~~~~~~~~~~~~
1 error generated.
```

Would you mind taking a look?  Thanks!

**Comment by kazutakahirata** [comment]
I've put in a workaround: e1fdaaafc51bf7323813139b2809ffea43e65ee4.

**Comment by anjenner** [comment]
> I've put in a workaround: [e1fdaaa](https://github.com/llvm/llvm-project/commit/e1fdaaafc51bf7323813139b2809ffea43e65ee4).

Thanks, Kazu. That looks fine for now but the proper fix is part of https://github.com/llvm/llvm-project/pull/105553 .
